### PR TITLE
Nationality code optimisation

### PR DESCRIPTION
**Pull request by abidfs**
### Change description
Added common component for nationality page
Consumed nationality component in `applicant1/nationality` and `children/nationality`
Created common get and post controllers for nationality
Consumed common nationality get and post controllers in `applicant1`, `applicant2` and `children`

### JIRA link

**Does this PR introduce a breaking change?** (check one with "x")

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/main/steps/common/components/nationality.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                addButton: {
+                  type: 'button',
+                  label: l => l.add,
+                  classes: 'govuk-button--secondary',
+                  value: 'addButton',
+                },
+              }),
+        },
+      },
+      ...(![FieldPrefix.APPLICANT1, FieldPrefix.APPLICANT1].includes(fieldPrefix)
```
Comment (renJosephHMCTS) [inline src/main/steps/common/components/nationality.ts:140]: Please can you explain this logic? It looks like a possible mistake

File: src/main/steps/children/nationality/content.ts [hunk 2]
```diff
@@ -1,158 +1,54 @@
...
-  british: 'British',
-  britishSubtext: 'including English, Scottish, Welsh and Northern Irish',
-  irish: 'Irish',
-  differentCountry: 'Citizen of a different country',
-  childrenNationality: 'Country name',
-  add: 'Add',
-  another: 'Add another country',
-  or: 'or',
-  notSure: 'Not Sure',
   errors: {
```
Comment (renJosephHMCTS) [inline src/main/steps/children/nationality/content.ts:14]: I believe this is already included in the common component so error messages may be removed in the content.ts implementations

File: src/main/steps/common/components/nationality.ts (hunk 1 above)
Comment (abidfs) [reply src/main/steps/common/components/nationality.ts:140]: Applicant1 and Applicant2 must be british residents hence notSure option is not applicable for them

File: src/main/steps/common/components/nationality.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+  [FieldPrefix.CHILDREN]: CHILDREN_NATIONALITY,
+};
+
+export const nationalityFields = (userCase: Partial<Case>, fieldPrefix: FieldPrefix): FormFields => ({
+  [`${fieldPrefix}Nationality`]: {
+    type: 'checkboxes',
+    label: l => l.label,
+    labelSize: 'l',
+    hint: l => l.hint,
+    // validator: atLeastOneFieldIsChecked,
```
[the remark this answers is not in the document]
Comment (abidfs) [reply src/main/steps/common/components/nationality.ts:67]: my bad.. removed the commented code now

File: src/main/steps/common/components/nationality.test.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+      required: 'Select if you are British, Irish or a citizen of a different country (in Welsh)',
+      notSureViolation: "Select a nationality or 'Not sure' (in Welsh)",
+    },
+    addAnotherNationality: {
+      required: 'This is not a valid entry (in Welsh)',
+    },
+  },
+};
+
+const langAssertions = (language, content) => {
```
[the remark this answers is not in the document]
Comment (abidfs) [reply src/main/steps/common/components/nationality.test.ts:61]: That's a great one.. thanks for adding it
Can we revisit this after your PR has merged

File: src/main/app/nationality/NationalityPostController.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    Object.assign(req.session.userCase, formData);
+
+    if (req.session.errors.length === 0) {
+      if (formData.addButton) {
+        if (!req.session.userCase[`${this.fieldPrefix}AdditionalNationalities`]) {
+          req.session.userCase[`${this.fieldPrefix}AdditionalNationalities`] = [];
+        }
+        if (formData.addAnotherNationality) {
+          req.session.userCase[`${this.fieldPrefix}AdditionalNationalities`]?.push(formData.addAnotherNationality);
+          req.session.userCase.addAnotherNationality = '';
```
[the remark this answers is not in the document]
Comment (abidfs) [reply src/main/app/nationality/NationalityPostController.ts:32]: We are resetting the value of `addAnotherNationality` on line 32 as soon as we are done using it (added it to the additionalNationalities array)
Can you think of any specific scenario?

File: src/main/steps/common/components/nationality.ts (hunk 1 above)
Comment (renJosephHMCTS) [reply src/main/steps/common/components/nationality.ts:139]: In that case, should the array contain FieldPrefix.APPLICANT2?

File: src/main/app/nationality/NationalityPostController.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (renJosephHMCTS) [reply src/main/app/nationality/NationalityPostController.ts:32]: Scenario:
User is doing birth mother flow
Enters text in the addAnotherNationality field
Exits the page without submitting
Starts the birth father flow
Nationality page is pre-populated with data

Would this happen, or would the post controller be called when exiting the page anyway?

File: src/main/steps/common/components/nationality.test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (renJosephHMCTS) [reply src/main/steps/common/components/nationality.test.ts:61]: Sure

File: src/main/steps/common/components/nationality.ts (hunk 1 above)
Comment (abidfs) [reply src/main/steps/common/components/nationality.ts:139]: 🙉 indeed yes

File: src/main/app/nationality/NationalityPostController.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (abidfs) [reply src/main/app/nationality/NationalityPostController.ts:32]: No this scenario would not happen, if user exits the page, then `addAnotherNationality` is never passed to server so we are safe

**Review by renJosephHMCTS — APPROVED** [review]

**Pull request merged**